File: android-app/app/src/main/java/com/atlasmasa/android/domain/LocalReasoningEngine.kt
package com.atlasmasa.android.domain

import java.util.Locale

class LocalReasoningEngine {
    data class Output(
        val summary: String,
        val nextAction: String,
        val confidence: Double,
        val model: String = "atlas-mobile-local-reasoner-v1",
    )

    fun reason(prompt: String, notesContext: List<UserNote>): Output {
        val compact = prompt.trim().take(420)
        if (compact.isEmpty()) {
            return Output(
                summary = "No prompt supplied.",
                nextAction = "Add one clear prompt and rerun queue.",
                confidence = 0.0,
            )
        }
        val lower = compact.lowercase(Locale.getDefault())
        val isHebrew = compact.any { it.code in 0x0590..0x05FF }

        val emergencySignals = listOf("emergency", "critical", "bleeding", "unconscious", "חירום", "דחוף", "דימום")
        if (emergencySignals.any { lower.contains(it) }) {
            return if (isHebrew) {
                Output(
                    summary = "מצב חירום זוהה. עדיפות מיידית לבטיחות, טריאז' ותקשורת רציפה.",
                    nextAction = "בצעו עכשיו: 1) אבטחת זירה 2) טריאז' 3) יצירת קשר עם חירום 4) שיתוף מיקום 5) תיעוד זמנים.",
                    confidence = 0.97,
                )
            } else {
                Output(
                    summary = "Emergency context detected. Prioritize scene safety, triage, and communications continuity.",
                    nextAction = "Do now: 1) secure scene 2) triage 3) contact emergency/satellite 4) share exact location 5) log timeline.",
                    confidence = 0.97,
                )
            }
        }

        val wealthSignals = listOf("salary", "job", "income", "revenue", "sales", "promotion", "עסק", "הכנסה", "משכורת")
        if (wealthSignals.any { lower.contains(it) }) {
            return if (isHebrew) {
                Output(
                    summary = "מסלול צמיחה כלכלית זוהה. המיקוד הוא מהלך הכנסה קצר-טווח עם מנגנון צמיחה חוזר.",
                    nextAction = "בחרו מסלול אחד ל-14 יום: שדרוג שכר, לקוח ראשון, או שיפור הצעת ערך. קבעו KPI יומי ובקרה שבועית.",
                    confidence = 0.91,
                )
            } else {
                Output(
                    summary = "Wealth-growth context detected. Focus on one near-term income move with a repeatable growth loop.",
                    nextAction = "Pick one 14-day route: raise comp, land first client, or upgrade offer. Track daily KPI + weekly review.",
                    confidence = 0.91,
                )
            }
        }

        val noteCount = notesContext.size.coerceAtMost(4)
        return if (isHebrew) {
            Output(
                summary = "תדריך ביצוע: ${compact.take(170)}${if (compact.length > 170) "..." else ""} | פתקי הקשר: $noteCount",
                nextAction = "ב-15 הדקות הקרובות: בצעו צעד אחד מדיד והוסיפו תיעוד תוצאה.",
                confidence = 0.78,
            )
        } else {
            Output(
                summary = "Execution brief: ${compact.take(170)}${if (compact.length > 170) "..." else ""} | Context notes: $noteCount",
                nextAction = "In the next 15 minutes: execute one measurable step and log outcome.",
                confidence = 0.78,
            )
        }
    }
}

File: windows-app/AtlasMasaWindows/Services/LocalReasoningEngine.cs
using AtlasMasaWindows.Models;
using System.Text.RegularExpressions;

namespace AtlasMasaWindows.Services;

public sealed class LocalReasoningEngine
{
    private readonly RustReasoningBridge _rustBridge = new();

    private static readonly string[] EmergencySignals =
    [
        "emergency", "urgent", "critical", "bleeding", "unconscious", "collapse",
        "חירום", "דחוף", "דימום", "חסר הכרה"
    ];

    private static readonly string[] WealthSignals =
    [
        "income", "salary", "promotion", "revenue", "sales", "client", "business",
        "הכנסה", "משכורת", "קידום", "לקוחות", "עסק"
    ];

    public string RuntimeStatusLine => _rustBridge.StatusLine;

    public async Task<LocalReasoningOutput> ReasonAsync(
        string prompt,
        IReadOnlyList<NoteRecord> notes,
        SystemPerformanceProfile perf,
        Action<double, string>? progressCallback = null,
        CancellationToken cancellationToken = default)
    {
        var normalized = prompt.Trim();
        if (string.IsNullOrWhiteSpace(normalized))
        {
            return BuildOutput(
                "atlas-windows-local-reasoner-v1",
                "No prompt supplied.",
                "Add one clear objective and queue it again.",
                0.0,
                "No reasoning could be performed because the prompt was empty.",
                ["Wait without setting an objective.", "Ask for more context before defining any target."],
                ["A clear user objective is required before planning can begin."]
            );
        }

        progressCallback?.Invoke(0.10, "Rust local reasoner");
        var rustOutput = await _rustBridge.TryReasonAsync(normalized, notes.Count, cancellationToken);
        if (rustOutput is not null)
        {
            progressCallback?.Invoke(1.0, "Completed via Rust");
            return rustOutput;
        }

        var isHebrew = Regex.IsMatch(normalized, @"[\u0590-\u05FF]");
        var lower = normalized.ToLowerInvariant();
        progressCallback?.Invoke(0.15, "Parsing prompt");
        await Task.Delay(perf.HighPerformanceMode ? 40 : 120, cancellationToken);

        if (EmergencySignals.Any(lower.Contains))
        {
            progressCallback?.Invoke(0.70, "Emergency protocol route");
            await Task.Delay(perf.HighPerformanceMode ? 80 : 180, cancellationToken);
            return isHebrew
                ? BuildOutput(
                    "atlas-windows-local-reasoner-v1",
                    "זוהה מצב חירום. נדרש תעדוף בטיחות, טריאז' ושרידות תפעולית.",
                    "בצעו עכשיו: אבטחת זירה, קריאה לחירום, שיתוף מיקום, ותיעוד זמנים.",
                    0.97,
                    "זוהו אותות חירום, ולכן נבחר מסלול שמעדיף בטיחות ותגובה מיידית על פני ניתוח רחב.",
                    ["להעמיק באבחון לפני פעולה.", "להמתין לעוד פרטים לפני התחלת תגובת חירום."],
                    ["ייתכן שקיים סיכון מיידי.", "פעולה מהירה חשובה כרגע יותר מניתוח ארוך."]
                )
                : BuildOutput(
                    "atlas-windows-local-reasoner-v1",
                    "Emergency context detected. Prioritize safety, triage, and continuity.",
                    "Do now: secure scene, contact emergency services, share location, log timeline.",
                    0.97,
                    "Emergency signals were present, so the response prioritized immediate safety and continuity over broader analysis.",
                    ["Pause for more context before acting.", "Start with a broader strategic plan first."],
                    ["There may be immediate risk or instability.", "Fast action matters more than exhaustive analysis right now."]
                );
        }

        if (WealthSignals.Any(lower.Contains))
        {
            progressCallback?.Invoke(0.70, "Wealth route synthesis");
            await Task.Delay(perf.HighPerformanceMode ? 80 : 180, cancellationToken);
            return isHebrew
                ? BuildOutput(
                    "atlas-windows-local-reasoner-v1",
                    "זוהה הקשר צמיחה כלכלית. נדרש מהלך הכנסה מדיד עם לולאת שיפור.",
                    "בחרו מהלך 14 יום אחד: קידום שכר, לקוח ראשון, או שדרוג הצעת ערך + KPI יומי.",
                    0.91,
                    "זוהה הקשר כלכלי, ולכן נבחר מסלול הכנסה יחיד ומדיד כדי לצמצם פיזור ולחזק מומנטום.",
                    ["לרדוף אחרי כמה מסלולים במקביל.", "להישאר בשלב למידה ארוך יותר לפני בחירה."],
                    ["מיקוד במסלול אחד יפיק מומנטום מהר יותר.", "לולאות מדידה קצרות עדיפות כרגע על תכנית מורכבת."]
                )
                : BuildOutput(
                    "atlas-windows-local-reasoner-v1",
                    "Wealth-growth context detected. Focus on one measurable income route.",
                    "Select one 14-day route: compensation upgrade, first client, or offer improvement with daily KPI.",
                    0.91,
                    "The prompt pointed to income growth, so the response chose one measurable route to reduce diffusion and improve execution speed.",
                    ["Pursue multiple income strategies in parallel.", "Stay in planning mode longer before committing."],
                    ["A narrower execution path will outperform a scattered one right now.", "Short feedback loops matter more than complexity here."]
                );
        }

        progressCallback?.Invoke(0.65, "Execution synthesis");
        await Task.Delay(perf.HighPerformanceMode ? 70 : 140, cancellationToken);

        var firstSentence = normalized
            .Split(['.', '?', '!', '\n'], StringSplitOptions.RemoveEmptyEntries)
            .FirstOrDefault()?
            .Trim() ?? normalized;
        if (firstSentence.Length > 180)
        {
            firstSentence = $"{firstSentence[..180]}...";
        }

        var notesUsed = Math.Min(4, notes.Count);
        progressCallback?.Invoke(1.0, "Completed");
        return isHebrew
            ? BuildOutput(
                "atlas-windows-local-reasoner-v1",
                $"תדריך ביצוע: {firstSentence} | פתקי הקשר: {notesUsed}",
                "ב-15 הדקות הקרובות: בצעו צעד אחד מדיד ורשמו תוצאה.",
                0.79,
                "נבחר מסלול פעולה ישיר לאחר השוואה מהירה בין חלופות, עם עדיפות לביצוע מיידי.",
                ["להעמיק באבחון לפני המלצה.", "לתת תכנית רחבה יותר ופחות דחופה."],
                [$"נעשה שימוש ב-{notesUsed} פתקי הקשר רלוונטיים.", "נדרשת התקדמות מעשית יותר מאשר ניתוח ארוך כרגע."]
            )
            : BuildOutput(
                "atlas-windows-local-reasoner-v1",
                $"Execution brief: {firstSentence} | Context notes used: {notesUsed}",
                "In the next 15 minutes: execute one measurable step and log the result.",
                0.79,
                "A direct execution path was chosen after a quick comparison of alternatives, prioritizing speed, clarity, and follow-through.",
                ["Ask more diagnostic questions before acting.", "Offer a broader and slower multi-step plan."],
                [$"Used {notesUsed} relevant context notes.", "A practical next step is more valuable than deeper analysis right now."]
            );
    }

    private static LocalReasoningOutput BuildOutput(
        string model,
        string summary,
        string nextAction,
        double confidence,
        string reasoningSummary,
        IReadOnlyList<string> alternatives,
        IReadOnlyList<string> assumptions)
    {
        return new LocalReasoningOutput
        {
            Model = model,
            Summary = summary,
            NextAction = nextAction,
            Confidence = confidence,
            GeneratedAt = DateTimeOffset.UtcNow,
            ReasoningSummary = reasoningSummary,
            AlternativesConsidered = alternatives.Where(item => !string.IsNullOrWhiteSpace(item)).ToList(),
            Assumptions = assumptions.Where(item => !string.IsNullOrWhiteSpace(item)).ToList(),
            ConfidenceLabel = ConfidenceLabel(confidence)
        };
    }

    private static string ConfidenceLabel(double confidence) => confidence switch
    {
        < 0.45 => "Low",
        < 0.72 => "Medium",
        < 0.9 => "High",
        _ => "Very High"
    };
}
